src/main/java/seedu/duke/parser/Parser.java

```diff
@@ -263,10 +269,59 @@ public static void parseEditStep(RecipeList recipeList,
             throw new Exception("error in edit step:\n" + e.getMessage());
         }
     }
-    public static String removeForbiddenChars(String ingredient) {
-        for (String chara : StringLib.FORBIDDEN_CHARS) {
-            ingredient.replaceAll(chara, "");
+    public static String[] parseAddToRecipeDescription(String description) {
+        String[] subDescriptions = description.split("/", 3);
+        String element0 = subDescriptions[0].toLowerCase().split("id")[0].trim();
+        subDescriptions[0] = element0.replace("-","");
+        String element1 = subDescriptions[1].toLowerCase().split("desc")[0].trim();
+        subDescriptions[1] = element1;
+        return subDescriptions;
+    }
+    public static String[] parseDeleteFromRecipeDescription(String description) {
+        String[] subDescriptions = description.split("/", 2);
+        String element0 = subDescriptions[0].toLowerCase().split("id")[0].trim();
+        subDescriptions[0] = element0.replace("-","");
+        return subDescriptions;
+    }
+    public static boolean isValidAddToRecipe(String description) {
+        String descLowerCase = description.toLowerCase().trim();
+        if ((descLowerCase.contains("step") ||
+                descLowerCase.contains("ingredient")) &&
+                descLowerCase.contains("id/") &&
+                descLowerCase.contains("desc/") &&
+                !(descLowerCase.contains("step") &&
+                        descLowerCase.contains("ingredient"))) {
+            return true;
+        } else {
+            return false;
+        }
+    }
+    public static boolean isValidDeleteFromRecipe(String description) {
+        String descLowerCase = description.toLowerCase().trim();
+        if ((descLowerCase.contains("step") ||
+                descLowerCase.contains("ingredient")) &&
+                descLowerCase.contains("id/") &&
+                !(descLowerCase.contains("step") &&
+                        descLowerCase.contains("ingredient"))) {
+            return true;
+        } else {
+            return false;
+        }
+    }
+    public static boolean isDuplicateIngredient(IngredientList ingredientList, String newIngredient) {
```

Comment:
@YatPang Has the `newIngredient` been trimmed of any leading and trailing spaces? Perhaps can check on the `Command#excecute()` method before you go on to compare with the ingredients that are already in the list.

Reply:
Confirmed newIngredient has been trimmed when parsing user input. See Parser line 270 - 277.